intel_adsp: ace: secondary core context save and restore

tmleman opened the pull request:
Reusing primary core context save/restore flow for purpose of secondary core D0 -> D3 -> D0 transitions.

FW on secondary core when preparing for D3 will save its context. When core is re-enabled we use _**dsp_restore_vector**_ as the FW entry point.

tmleman committed:
intel_adsp: ace: add stack for other cores

This patch replace temporary stack of the restore vector with interrupt
stack to reduce memory usage. Additionally we can assign seprate stack
for each core. This will allow to reuse this vector for secondary cores.

Signed-off-by: Tomasz Leman <tomasz.m.leman@intel.com>

tmleman committed:
intel_adsp: ace: cpu context save refactor

This patch is preparing cpu context save and restore code so it can be
later used by the multiple cores.

Signed-off-by: Tomasz Leman <tomasz.m.leman@intel.com>

tmleman committed:
intel_adsp: ace: secondary core context restore

Reusing primary core context save/restore flow for purpose of secondary
core D0 -> D3 -> D0 transitions. If core is re-enabled we use
dsp_restore_vector as the FW entry point.

Signed-off-by: Tomasz Leman <tomasz.m.leman@intel.com>

andyross commented on soc/xtensa/intel_adsp/ace/power.c:
```diff
@@ -176,8 +188,11 @@ __asm__(".align 4\n\t"
+	"  movi  a3, 0x400\n\t"
+	"  mull  a2, a2, a3\n\t"
+	"  add   a2, a2, a3\n\t"
+	"  add   sp, sp, a2\n\t"
```
What's going on here?  You're computing ((PRID + 1) * 1024) and popping that many bytes off the stack?  It was weird enough before where there seems to have been a 4k buffer being adjusted, but I don't understand at all why the stack adjustment needs to be CPU-specific.  This is the CPU-local interrupt stack, isn't it?

andyross replied:
OK, now I get it.  This is just computing the end of the relevant subarray, and doing the computation at runtime because the __asm__ block is global and so can't feed C expressions via constraints.  That's annoying, I've been there.  But can you clean this up so you aren't hardcoding 0x400 == CONFIG_IDLE_STACK_SIZE, or at the very least put a comment here explaining what's going on?

Still, I don't get the point here.  Why are you wasting 4k (!!) of precious HPSRAM for per-core power stacks, when there's a perfectly good and completely unused (because interrupts are masked, right?  right?!) per-core interrupt stack already.

Please find a way to unify this with the interrupt stack.

andyross commented on soc/xtensa/intel_adsp/ace/power.c:
```diff
@@ -323,7 +334,7 @@ __weak void pm_state_exit_post_ops(enum pm_state state, uint8_t substate_id)
 		}
 #endif /* CONFIG_ADSP_IMR_CONTEXT_SAVE */
-
+		z_mark_thread_as_not_suspended(z_current_get());
```
This isn't right.  The "suspended" state means that a thread is neither running currently nor in the run queue[1].  By definition, an idle thread can never be blocked, and by implementation choice is never added to the run queue.  It just doesn't make sense here.

andyross commented on soc/xtensa/intel_adsp/ace/power.c:
```diff
@@ -222,9 +237,9 @@ __weak void pm_state_set(enum pm_state state, uint8_t substate_id)
-		soc_cpus_active[cpu] = false;
-		z_xtensa_cache_flush_inv_all();
 		if (cpu == 0) {
+			soc_cpus_active[cpu] = false;
```
Won't this now fail to update soc_cpus_active[] for secondare cores?

tmleman replied:
No. For secondary core its done inside of _power_gate_entry_ function.

andyross reviewed (changes requested):
Can we clean this up?  It seems like you needed a way to tell whether or not a core was being started up after initial boot, and decided to abuse the kernel thread state for the idle thread?  That's... no, that's just not right.  Those are different subsystems.  If the arch/soc code needs to track state for its own purposes, it needs to do it on its own and not repurpose other code from other areas.

FWIW: looking at that soc_cpus_active[] array, I'm guessing that's the right kind of abstraction to be using (though it seems to have been broken by this PR?).

tmleman commented:
@andyross please take a look at the updated versions now.

dcpleung commented on soc/xtensa/intel_adsp/ace/power.c:
```diff
@@ -163,9 +166,12 @@ __asm__(".align 4\n\t"
-	"  add sp, sp, a2\n\t"
+	"  movi  sp, z_interrupt_stacks\n\t"
+	"  rsr   a2, PRID\n\t"
+	"  movi  a3, 0x800\n\t"
```
Could you try `"movi a3, " CONFIG_ISR_STACK_SIZE "\n\t"`? This should embed the value and you can skip the BUILD_ASSERT above.

softwarecki replied:
`"movi a3, " STRINGIFY(CONFIG_ISR_STACK_SIZE)  "\n\t"` should work ;)

ceolin commented on soc/xtensa/intel_adsp/ace/multiprocessing.c:
```diff
@@ -7,6 +7,8 @@
 #include <zephyr/sys/check.h>
 #include <zephyr/arch/cpu.h>
+#include <zephyr/pm/pm.h>
+#include <ksched.h>
```
why are you including `ksched.h` here ?

tmleman replied:
Thanks for catching this. Those are leftovers from the previous version. I forgot to delete while editing.

ceolin commented on soc/xtensa/intel_adsp/ace/multiprocessing.c:
```diff
@@ -82,9 +84,21 @@ void soc_start_core(int cpu_num)
+		 * initialization, the next pm state is set to ACTIVE. This way we can determine
+		 * whether the core is being turned on again or for the first time.
+		 */
+		if (pm_state_next_get(cpu_num)->state == PM_STATE_ACTIVE) {
```
PM is optional, one can build without it. If you build wihout PM, `pm_state_next_get()` will return a NULL pointer. This change can cause a null pointer dereference. Please account this possibility here.

ceolin commented on soc/xtensa/intel_adsp/ace/power.c:
```diff
@@ -163,9 +178,12 @@ __asm__(".align 4\n\t"
-	"  movi  sp, d0i3_stack\n\t"
-	"  movi a2, 0x1000\n\t"
-	"  add sp, sp, a2\n\t"
+	"  movi  sp, z_interrupt_stacks\n\t"
```
isn't `a1` the `sp` ? It is just confuse to me mixing nomenclature

tmleman replied:
Yes, they are the same. But I don't see how it's confusing. In this part of code I'm calculating correct stack pointer.

ceolin commented on soc/xtensa/intel_adsp/ace/power.c:
```diff
@@ -134,27 +134,42 @@ void dsp_restore_vector(void);
-	struct lpsram_header *lpsheader =
-		(struct lpsram_header *) DT_REG_ADDR(DT_NODELABEL(sram1));
-
 	xthal_window_spill();
```
I particularly don't like using HAL API when we have equivalent implemented. We could use our API the limitation is that we need save `a0` and do `call0` in assembler.

tmleman replied:
Could you explain more precisely what you mean?

ceolin replied:
I'm just saying that I prefer using Zephyr's native API (when avaialable) than the HAL API. That's said the API to spill window register on Zephyr has constraints different from the one in the HAL and I was explaining that.

tmleman replied:
I prefer to keep it as it is. This PR is resolving issues with secondary cores power flows and I don't wanna do too much refactoring in it.

ceolin replied:
that is reasonabe

ceolin commented on soc/xtensa/intel_adsp/ace/power.c:
```diff
@@ -210,9 +228,9 @@ __weak void pm_state_set(enum pm_state state, uint8_t substate_id)
 		DSPCS.bootctl[cpu].bctl &= ~DSPBR_BCTL_WAITIPCG;
-		soc_cpus_active[cpu] = false;
-		z_xtensa_cache_flush_inv_all();
 		if (cpu == 0) {
```
Why did you move this if ? Don't you need `soc_cpus_active[cpu] = false;` for cores != 0 too ?

tmleman replied:
For secondary core its done in _**power_gate_entry**_ function.

ceolin reviewed (changes requested):
Giving a  -1 to avoid it be merged why there are open questions

ceolin commented:
> Yes, they are the same. But I don't see how it's confusing. In this part of code I'm calculating correct stack pointer. 

I know, but mixing names is confusing IMHO. I saw only one more place using `sp` for `a1`, everywhere else we are just using `a1`. This may be confuse for someone that is not super used to Xtensa.

ceolin commented on soc/xtensa/intel_adsp/ace/power.c:
```diff
@@ -210,9 +222,9 @@ __weak void pm_state_set(enum pm_state state, uint8_t substate_id)
 		z_xt_ints_off(0xffffffff);
```
Why do you need this here ? I'm looking these masks and I don't understand. This function enters will interruptions masked XCHAL_NMILEVEL. If you don't need to do it, you could move `core_desc[cpu].intenable = XTENSA_RSR("INTENABLE");` to `_save_core_context` and avoid unecessary instructions and duplicated code.

tmleman replied:
For primary core _save_core_context is done only when CONFIG_ADSP_IMR_CONTEXT_SAVE is enabled.

ceolin commented on soc/xtensa/intel_adsp/ace/power.c:
```diff
@@ -210,9 +222,9 @@ __weak void pm_state_set(enum pm_state state, uint8_t substate_id)
-		z_xtensa_cache_flush_inv_all();
 		if (cpu == 0) {
+			soc_cpus_active[cpu] = false;
+			z_xtensa_cache_flush_inv_all();
```
couldn't this be done inside `save_core_context` too ? Because you are doing this big cache invalidation and then do other times in this function for smaller memory chunks. If you put it in the end you could do this flush/invalidation only once.

tmleman replied:
In case of a secondary core I want to change this value only when everything is ready to turn it off.

ceolin commented:
Please take a look into my comments. They are not things added now, but since there is you are changing some parts of this code they make sense. This logic is getting overcomplicated and confusing IMHO

tmleman committed:
intel_adsp: power: ace: interrupts on and off

Masking out all interrupt during power state transition and restoring
them after is now common thing for all power states. No need to
duplicate code.

Signed-off-by: Tomasz Leman <tomasz.m.leman@intel.com>

tmleman committed:
intel_adsp: power: ace: cache invalidation

This patch moves cache flush/invalidation to section executed only when
IMR context saving is enabled. If this option is disabled no FW context
is stored so any lost data doesn't matter.

Signed-off-by: Tomasz Leman <tomasz.m.leman@intel.com>

tmleman commented:
> Please take a look into my comments. They are not things added now, but since there is you are changing some parts of this code they make sense. This logic is getting overcomplicated and confusing IMHO

@ceolin I have addressed your comments in some way, please see if this version suits you.

ceolin commented on soc/xtensa/intel_adsp/ace/power.c:
```diff
@@ -134,27 +133,37 @@ void dsp_restore_vector(void);
-		(struct lpsram_header *) DT_REG_ADDR(DT_NODELABEL(sram1));
-
 	xthal_window_spill();
+	z_xtensa_cache_flush_inv_all();
```
why don't you do this in the end of this function and avoid the following flushes ?

tmleman replied:
I need to do cache flush/invalidation for all after window spill and before context saving. It will not work otherwise.

ceolin commented on soc/xtensa/intel_adsp/ace/power.c:
```diff
@@ -134,27 +133,37 @@ void dsp_restore_vector(void);
+
+		lpsheader->adsp_lpsram_magic = LPSRAM_MAGIC_VALUE;
+		lpsheader->lp_restore_vector = &dsp_restore_vector;
+		z_xtensa_cache_flush(lpsheader, sizeof(struct lpsram_header));
```
here

ceolin commented on soc/xtensa/intel_adsp/ace/power.c:
```diff
@@ -134,27 +133,37 @@ void dsp_restore_vector(void);
 	soc_cpus_active[core_id] = false;
-	z_xtensa_cache_flush_inv_all();
-	z_xt_ints_on(ALL_USED_INT_LEVELS_MASK);
+	z_xtensa_cache_flush(soc_cpus_active, sizeof(soc_cpus_active));
```
and here

tmleman commented:
@andyross friendly reminder

tmleman commented:
Adding DNM label because of https://github.com/thesofproject/sof/issues/7433. After this PR is merged zephyr update on SOF side will require additional changes. This would make later bisect more difficult in case of any new problems.

ceolin reviewed (dismissed):
I'm still a little bit skeptical about the need of all flush/invalidate cache but you know better than me about it here

ceolin commented:
I'm still a little bit skeptical about the need of all flush/invalidate cache but you know better than me about it here

tmleman commented:
Rebase after merge of  https://github.com/zephyrproject-rtos/zephyr/pull/50136

andyross reviewed (approved):
Saw this come by again.  Looks great to me, my concerns earlier were all addressed.
